infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java

```diff
@@ -133,37 +141,54 @@
      * Initialization task.
      * This task is blocked until the management server is bootstrapped and has a connected runtime.
      * Performs recovery if required.
-     * Bootstraps the primary sequencer if the server has been bootstrapped.
-     * Initiates the failure detection and healing detection tasks.
+     * Initiates the monitoring services which performs failure detection and healing detection tasks.
      */
     private void initializationTask() {
         log.info("Start initialization task");
 
+        // Wait for management server bootstrapped.
         try {
             while (!shutdown && serverContext.getManagementLayout() == null) {
-                log.warn("Management Server waiting to be bootstrapped");
+                log.warn("initializationTask: Management Server waiting to be bootstrapped");
                 Sleep.MILLISECONDS.sleepRecoverably(CHECK_BOOTSTRAP_INTERVAL.toMillis());
             }
+        } catch (InterruptedException e) {
+            // Due to shutdown, which interrupts this thread
+            log.debug("initializationTask: Initialization task interrupted without " +
+                    "management server bootstrapped");
+            return;
+        }
 
-            if (shutdown){
-                log.info("Shutdown signal. Interrupt initialization");
+        // Retry layout recovery if need until success.
+        long recoveryAttempts = 0;
+        while (!shutdown && !recovered) {
```

Comment:
Are you sure that we don't break failure detector here?
It checked `serverContext.getManagementLayout() == null` in prev version, how we know now that managementLayout is not null?

Reply:
This check is still there in the above while block. I didn't remove it.

Reply:
ok